File: app/assets/javascripts/discover/DiscoverController.js
(function () {
    var app = angular.module('shocktrade');

    /**
     * Discover Controller
     * @author lawrence.daniels@gmail.com
     */
    app.controller('DiscoverController', ['$scope', '$cookieStore', '$interval', '$log', '$routeParams', '$timeout', 'toaster', 'FavoriteSymbols', 'HeldSecurities', 'MarketStatus', 'MySession', 'NewOrderDialog', 'QuoteService', 'RecentSymbols',
        function ($scope, $cookieStore, $interval, $log, $routeParams, $timeout, toaster, FavoriteSymbols, HeldSecurities, MarketStatus, MySession, NewOrderDialog, QuoteService, RecentSymbols) {

            // setup a private loading variable
            $scope.loading = false;

            // setup the public variables
            $scope.marketClock = (new Date()).toTimeString();
            $scope.ticker = null;
            $scope.q = {active: true};

            // define the display options
            $scope.options = {
                range: $cookieStore.get("chart_range") || "5d"
            };

            // setup the chart range
            $scope.$watch("options.range", function (newValue, oldValue) {
                $cookieStore.put("chart_range", newValue);
            });

            // setup filtered quotes & trading history
            $scope.filterQuotes = [];
            $scope.tradingHistory = null;
            $scope.selectedTradingHistory = null;

            // define the Quote module expanders
            $scope.expanders = [{
                title: "Performance & Risk",
                url: "/assets/views/discover/quotes/expanders/price_performance.htm",
                icon: "fa-line-chart",
                expanded: false,
                visible: function (q) {
                    return q.high52Week || q.low52Week || q.change52Week || q.movingAverage50Day ||
                        q.movingAverage200Day || q.change52WeekSNP500 || q.beta;
                }
            }, {
                title: "Income Statement",
                url: "/assets/views/discover/quotes/expanders/income_statement.htm",
                icon: "fa-money",
                expanded: false,
                visible: function (q) {
                    return q.revenue || q.revenuePerShare || q.revenueGrowthQuarterly || q.grossProfit ||
                        q.EBITDA || q.netIncomeAvailToCommon || q.dilutedEPS || q.earningsGrowthQuarterly;
                }
            }, {
                title: "Balance Sheet",
                url: "/assets/views/discover/quotes/expanders/balanace_sheet.htm",
                icon: "fa-calculator",
                expanded: false,
                visible: function (q) {
                    return q.totalCash || q.totalDebt || q.currentRatio || q.totalCashPerShare ||
                        q.totalDebtOverEquity || q.bookValuePerShare || q.returnOnAssets ||
                        q.profitMargin || q.mostRecentQuarterDate || q.returnOnEquity ||
                        q.operatingMargin || q.fiscalYearEndDate;
                }
            }, {
                title: "Valuation Measures",
                url: "/assets/views/discover/quotes/expanders/valuation_measures.htm",
                icon: "fa-gears",
                expanded: false,
                visible: function (q) {
                    return q.enterpriseValue || q.trailingPE || q.forwardPE || q.pegRatio || q.priceOverSales ||
                        q.priceOverBookValue || q.enterpriseValueOverRevenue || q.enterpriseValueOverEBITDA ||
                        q.operatingCashFlow || q.leveredFreeCashFlow;
                }
            }, {
                title: "Share Statistics",
                url: "/assets/views/discover/quotes/expanders/share_statistics.htm",
                icon: "fa-bar-chart",
                expanded: false,
                visible: function (q) {
                    return q.avgVolume3Month || q.avgVolume10Day || q.sharesOutstanding || q.sharesFloat ||
                        q.pctHeldByInsiders || q.pctHeldByInstitutions || q.sharesShort || q.shortRatio ||
                        q.shortPctOfFloat || q.sharesShortPriorMonth;
                }
            }, {
                title: "Dividends & Splits",
                url: "/assets/views/discover/quotes/expanders/dividends_splits.htm",
                icon: "fa-cut",
                expanded: false,
                visible: function (q) {
                    return q.forwardAnnualDividendRate || q.forwardAnnualDividendYield ||
                        q.trailingAnnualDividendYield || q.divYield5YearAvg || q.payoutRatio ||
                        q.dividendDate || q.exDividendDate || q.lastSplitFactor || q.lastSplitDate;
                }
            }, {
                title: "Historical Quotes",
                url: "/assets/views/discover/quotes/trading_history.htm",
                icon: "fa-calendar",
                expanded: false,
                loading: false,
                onExpand: function (callback) {
                    if ($scope.tradingHistory === null && $scope.q.assetType === 'Common Stock') {
                        $scope.loadTradingHistory($scope.q.symbol, callback);
                    }
                }
            }];

            $scope.autoCompleteSymbols = function (searchTerm) {
                return QuoteService.autoCompleteSymbols(searchTerm, 20)
                    .then(function (response) {
                        return response.data;
                    });
            };

            /**
             * Initializes the module
             */
            $scope.init = function () {
                // setup market status w/updates
                $interval(function () {
                    $scope.marketClock = (new Date()).toTimeString();
                }, 1000);

                // setup the market status updates
                setupMarketStatusUpdates();
            };

            $scope.popupNewOrderDialog = function (symbol) {
                NewOrderDialog.popup({symbol: symbol});
            };

            $scope.expandSection = function (module) {
                module.expanded = !module.expanded;
                if (module.expanded && module.onExpand) {
                    module.loading = true;
                    var promise = $timeout(function () {
                        module.loading = false;
                    }, 3000);
                    module.onExpand(function () {
                        module.loading = false;
                        $timeout.cancel(promise);
                    });
                }
            };

            $scope.addFavoriteSymbol = function (symbol) {
                FavoriteSymbols.add(symbol);
            };

            $scope.isFavorite = function (symbol) {
                return FavoriteSymbols.isFavorite(symbol);
            };

            $scope.removeFavoriteSymbol = function (symbol) {
                FavoriteSymbols.remove(symbol);
            };

            $scope.hasHoldings = function (q) {
                return q && q.products && (q.legalType === 'ETF') && (q.products.length !== 0);
            };

            $scope.removeRecentSymbol = function (symbol) {
                RecentSymbols.remove(symbol);
            };

            $scope.rowClass = function (column, row) {
                return ( column === "symbol" ) ? row['exchange'] : column;
            };

            $scope.columnAlign = function (column) {
                return ( column === "symbol" ) ? "left" : "right";
            };

            $scope.isOpen = function (filter) {
                return (filter.name === "Favorites");
            };

            $scope.getAssetIcon = function (quote) {
                var q = quote || $scope.q;
                if (q && q.assetType) {
                    switch (q.assetType) {
                        case 'Crypto-Currency' :
                            return "/assets/images/asset_types/bitcoin.png";
                        case 'Currency':
                            return "/assets/images/asset_types/currency.png";
                        case 'ETF':
                            return "/assets/images/asset_types/etf.png";
                        default:
                            return "/assets/images/asset_types/stock.png";
                    }
                }
                else return "/assets/images/status/transparent.png";
            };

            $scope.getMatchedAssetIcon = function (q) {
                //console.log("q = " + angular.toJson(q));
                return "/assets/images/asset_types/stock.png";
            };

            $scope.getRiskClass = function (riskLevel) {
                return riskLevel ? "risk_" + riskLevel.toLowerCase() : null;
            };

            $scope.getRiskDescription = function (riskLevel) {
                if (riskLevel === "Low") return "Generally recommended for investment";
                else if (riskLevel === "Medium") return "Not recommended for inexperienced investors";
                else if (riskLevel === "High") return "Not recommended for investment";
                else if (riskLevel === "Unknown") return "The risk level could not be determined";
                else return "The risk level could not be determined";
            };

            $scope.getBetaClass = function (beta) {
                if (beta == null) return "";
                else if (beta > 1.3 || beta < -1.3) return "volatile_red";
                else if (beta >= 0.0) return "volatile_green";
                else if (beta < 0) return "volatile_yellow";
                else return "";
            };

            $scope.loadTickerQuote = function (_ticker) {
                var ticker = $("#stockTicker").val() || _ticker;
                $scope.loadQuote(ticker);
            };

            $scope.loadQuote = function (ticker) {
                $log.info("Loading symbol " + angular.toJson(ticker));

                // setup the loading animation
                $scope.startLoading();

                // determine the symbol
                var symbol = null;
                if (ticker.symbol) {
                    symbol = ticker.symbol.toUpperCase();
                }
                else {
                    var index = ticker.indexOf(' ');
                    symbol = (index == -1 ? ticker : ticker.substring(0, index)).toUpperCase();
                }

                // load the quote
                QuoteService.loadStockQuote(symbol).then(
                    function (response) {
                        var quote = response.data;
                        if (quote) {
                            // capture the quote
                            $scope.q = quote;
                            $scope.ticker = quote.symbol + " - " + quote.name;

                            // save the cookie
                            $cookieStore.put('symbol', quote.symbol);

                            // add the symbol to the Recently-viewed Symbols
                            RecentSymbols.add(symbol);

                            // get the risk level
                            QuoteService.getRiskLevel(symbol).then(
                                function (response) {
                                    quote.riskLevel = response.data;
                                },
                                function (response) {
                                    toaster.pop('error', 'Error!', "Error retrieving risk level for " + symbol);
                                });

                            // load the trading history
                            $scope.tradingHistory = null;
                            if ($scope.expanders[6].expanded) {
                                $scope.expandSection($scope.expanders[6]);
                            }
                        }
                        else {
                            toaster.pop('error', 'Error!', "No quote found for " + symbol);
                            console.log("Empty quote? " + angular.toJson(quote));
                        }

                        // disabling the loading status
                        $scope.stopLoading();
                    },
                    function (response) {
                        $log.error("Failed to retrieve quote: " + response.status);
                        $scope.stopLoading();
                        toaster.pop('error', 'Error!', "Error loading quote " + symbol);
                    });
            };

            $scope.loadTradingHistory = function (symbol, callback) {
                QuoteService.getTradingHistory(symbol).then(
                    function (results) {
                        $scope.tradingHistory = results;
                        if (callback) callback();
                    },
                    function (response) {
                        toaster.pop('error', 'Error!', "Error loading trading history for " + symbol);
                        if (callback) callback();
                    });
            };

            $scope.loadFilterQuotes = function (filter, index) {
                $scope.loading = true;
                QuoteService.getFilterQuotes(filter).then(
                    function (quotes) {
                        $scope.loading = false;
                        filter.rows = quotes;
                    },
                    function (err) {
                        $scope.loading = false;
                        toaster.pop('error', 'Error!', "Error loading filter " + filter.name);
                    });
            };

            $scope.tradingActive = function (time) {
                return (new Date()).getTime();
            };

            $scope.selectTradingHistory = function(t) {
                $scope.selectedTradingHistory = t;
            };

            $scope.hasSelectedTradingHistory = function() {
                return $scope.selectedTradingHistory != null;
            };

            $scope.isSelectedTradingHistory = function(t) {
                return $scope.selectedTradingHistory === t;
            };

            // maintain the appropriate aspect ratio for the application
            var lastWindowUpdate = 0;
            var mainWidthPctValue = '55%';
            $scope.mainWidthPct = function () {
                var now = (new Date()).getTime();
                if (now - lastWindowUpdate >= 1000) {
                    lastWindowUpdate = now;
                    var width = $('body').innerWidth() - 20;
                    var adjWidth = ((1.0 - 459.0 / width) * 100.0).toFixed(0);
                    mainWidthPctValue = adjWidth + "%";
                    //console.log("pct = " + pct + ", adjWidth = " + adjWidth);
                }
                return mainWidthPctValue;
            };

            // watch for changes to the player's profile
            $scope.$watch("MySession.userProfile", function () {
                if (!MySession.userProfile.favorites) MySession.userProfile.favorites = ['AAPL'];
                if (!MySession.userProfile.recentSymbols) MySession.userProfile.recentSymbols = ['AAPL', 'AMZN', 'GOOG', 'MSFT'];

                // load the favorite and recent quotes
                FavoriteSymbols.setSymbols(MySession.userProfile.favorites);
                RecentSymbols.setSymbols(MySession.userProfile.recentSymbols);

                // setup the filters
                var filters = MySession.userProfile.filters;
                $scope.filterQuotes = [];
                for (var n = 0; n < filters.length; n++) {
                    var clonedFilter = $scope.clone(filters[n]);
                    clonedFilter.rows = [];
                    clonedFilter.maxResults = 10;
                    $scope.filterQuotes.push(clonedFilter);
                }

                // load the held securities
                var id = MySession.userProfile.OID();
                if (id) {
                    HeldSecurities.init(id);
                }
            });

            function setupMarketStatusUpdates() {
                $scope.usMarketsOpen = null;
                $log.info("Retrieving market status...");
                MarketStatus.getMarketStatus(function (response) {
                    // retrieve the delay in milliseconds from the server
                    var delay = response.delay;
                    if (delay < 0) {
                        delay = response.end - response.sysTime;
                        if (delay <= 300000) {
                            delay = 300000; // 5 minutes
                        }
                    }

                    // set the market status
                    $log.info("US Markets are " + (response.active ? 'Open' : 'Closed') + "; Waiting for " + delay + " msec until next trading start...");
                    setTimeout(function () {
                        $scope.usMarketsOpen = response.active;
                    }, 750);

                    // wait for the delay, then call recursively
                    setTimeout(function () {
                        setupMarketStatusUpdates();
                    }, delay);
                });
            }

            ///////////////////////////////////////////////////////////////////////////
            //          Initialization
            ///////////////////////////////////////////////////////////////////////////

            // load the symbol
            (function () {
                if (!$scope.q.symbol) {
                    // get the symbol
                    var symbol = $routeParams.symbol || $cookieStore.get('symbol') || RecentSymbols.getLast();

                    // load the symbol
                    $scope.loadQuote(symbol);
                }
            })();

        }]);

})();

File: app/client/angularjs/src/main/scala/com/shocktrade/stockguru/discover/DiscoverController.scala
package com.shocktrade.stockguru.discover

import com.shocktrade.common.models.Profile
import com.shocktrade.common.models.quote.{AutoCompleteQuote, CompleteQuote}
import com.shocktrade.stockguru._
import com.shocktrade.stockguru.dialogs.NewOrderDialog
import com.shocktrade.stockguru.dialogs.NewOrderDialogController.{NewOrderDialogResult, NewOrderParams}
import com.shocktrade.stockguru.discover.DiscoverController._
import com.shocktrade.stockguru.profile.ProfileService
import org.scalajs.angularjs.AngularJsHelper._
import org.scalajs.angularjs.cookies.Cookies
import org.scalajs.angularjs.http.HttpResponse
import org.scalajs.angularjs.toaster.Toaster
import org.scalajs.angularjs.{angular, injected, _}
import org.scalajs.dom.browser.console
import org.scalajs.nodejs.util.ScalaJsHelper._
import org.scalajs.sjs.JsUnderOrHelper._

import scala.concurrent.duration._
import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue
import scala.scalajs.js
import scala.scalajs.js.JSConverters._
import scala.scalajs.js.annotation.ScalaJSDefined
import scala.util.{Failure, Success, Try}

/**
  * Discover Controller
  * @author Lawrence Daniels <lawrence.daniels@gmail.com>
  */
class DiscoverController($scope: DiscoverControllerScope, $cookies: Cookies, $location: Location, $q: Q,
                         $routeParams: DiscoverRouteParams, $timeout: Timeout, toaster: Toaster,
                         @injected("MarketStatus") marketStatus: MarketStatusService,
                         @injected("MySessionService") mySession: MySessionService,
                         @injected("NewOrderDialog") newOrderDialog: NewOrderDialog,
                         @injected("ProfileService") profileService: ProfileService,
                         @injected("QuoteService") quoteService: QuoteService)
  extends AutoCompletionController($scope, $q, quoteService) with GlobalLoading {

  private var usMarketStatus: Either[MarketStatus, Boolean] = Right(false)

  // setup the public variables
  $scope.ticker = null
  $scope.q = CompleteQuote()

  // define the display options
  $scope.options = new DiscoverOptions(range = $cookies.getOrElse("chart_range", "5d"))

  // define the quote section expanders
  $scope.expanders = js.Array(
    new ModuleExpander(title = "Performance & Risk",
      url = "/views/discover/expanders/price_performance.html",
      icon = "fa-line-chart",
      visible = isPerformanceRisk),
    new ModuleExpander(title = "Income Statement",
      url = "/views/discover/expanders/income_statement.html",
      icon = "fa-money",
      visible = isIncomeStatement),
    new ModuleExpander(title = "Balance Sheet",
      url = "/views/discover/expanders/balance_sheet.html",
      icon = "fa-calculator",
      visible = isBalanceSheet),
    new ModuleExpander(title = "Valuation Measures",
      url = "/views/discover/expanders/valuation_measures.html",
      icon = "fa-gears",
      visible = isValuationMeasures),
    new ModuleExpander(title = "Share Statistics",
      url = "/views/discover/expanders/share_statistics.html",
      icon = "fa-bar-chart",
      visible = isShareStatistics),
    new ModuleExpander(title = "Dividends & Splits",
      url = "/views/discover/expanders/dividends_splits.html",
      icon = "fa-cut",
      visible = isDividendsSplits),
    new ModuleExpander(title = "Historical Quotes",
      url = "/views/discover/quotes/trading_history.html",
      icon = "fa-calendar")
  )

  ///////////////////////////////////////////////////////////////////////////
  //          Public Function
  ///////////////////////////////////////////////////////////////////////////

  $scope.expandSection = (aModule: js.UndefOr[ModuleExpander]) => aModule foreach { module =>
    module.expanded = !module.expanded
  }

  $scope.popupNewOrderDialog = (aSymbol: js.UndefOr[String]) => aSymbol map { symbol =>
    newOrderDialog.popup(new NewOrderParams(symbol = symbol))
  }

  $scope.loadQuote = (aValue: js.UndefOr[Any]) => aValue foreach {
    case ticker: String =>
      updateQuote(ticker.indexOf(" ") match {
        case -1 => ticker
        case index => ticker.substring(0, index)
      })

    case unknown =>
      Try(unknown.asInstanceOf[AutoCompleteQuote]) match {
        case Success(quote) =>
          console.log(s"Loading symbol from ${angular.toJson(quote, pretty = false)}")
          quote.symbol foreach updateQuote
        case Failure(e) =>
          console.error(s"Unhandled value '$unknown': ${e.displayMessage}")
      }
  }

  private def updateQuote(ticker: String) {
    // get the symbol (e.g. "AAPL - Apple Inc")
    val symbol = if (ticker.contains(" ")) ticker.substring(0, ticker.indexOf(" ")).trim else ticker

    // load the quote
    asyncLoading($scope)(quoteService.getCompleteQuote(symbol)) onComplete {
      case Success(quote) if quote.symbol.isAssigned =>
        $scope.$apply { () =>
          // capture the quote
          $scope.q = quote
          $scope.ticker = s"${quote.symbol} - ${quote.name}"

          // update the address bar
          $location.search("symbol", quote.symbol)

          // store the last symbol
          $cookies.put(LastSymbolCookie, quote.symbol)

          // add the symbol to the Recently-viewed Symbols
          mySession.addRecentSymbol(symbol)

          // load the trading history
          $scope.tradingHistory = null
          if ($scope.expanders(6).expanded) {
            $scope.expandSection($scope.expanders(6))
          }
        }
      case Success(quote) =>
        console.log(s"quote = ${angular.toJson(quote)}")
        toaster.warning(s"Symbol not found")

      case Failure(e) =>
        console.error(s"Failed to retrieve quote: ${e.getMessage}")
        toaster.error(s"Error loading quote $symbol")
    }
  }


  //////////////////////////////////////////////////////////////////////
  //              Type-Ahead Functions
  //////////////////////////////////////////////////////////////////////

  $scope.formatSearchResult = (aResult: js.UndefOr[AutoCompleteQuote]) => {
    for {
      result <- aResult.flat
      symbol <- result.symbol
    } yield symbol
  }

  $scope.onSelectedItem = (aItem: js.UndefOr[AutoCompleteQuote], aModel: js.UndefOr[AutoCompleteQuote], aLabel: js.UndefOr[String]) => {
    aModel.flatMap(_.symbol) foreach { symbol =>
      console.log(s"Loading '$symbol' => ${angular.toJson(aModel)}")
      $scope.loadQuote(symbol)
    }
  }

  ///////////////////////////////////////////////////////////////////////////
  //          Symbols - Favorites
  ///////////////////////////////////////////////////////////////////////////

  $scope.addFavoriteSymbol = (aSymbol: js.UndefOr[String]) => {
    for {
      symbol <- aSymbol
      userId <- mySession.userProfile._id
    } yield profileService.addFavoriteSymbol(userId, symbol)
  }

  $scope.isFavorite = (aSymbol: js.UndefOr[String]) => aSymbol.exists(mySession.isFavoriteSymbol)

  $scope.removeFavoriteSymbol = (aSymbol: js.UndefOr[String]) => {
    for {
      symbol <- aSymbol
      userId <- mySession.userProfile._id
    } yield profileService.removeFavoriteSymbol(userId, symbol)
  }

  ///////////////////////////////////////////////////////////////////////////
  //          Symbols - Recently Viewed
  ///////////////////////////////////////////////////////////////////////////

  $scope.addRecentSymbol = (aSymbol: js.UndefOr[String]) => {
    for {
      symbol <- aSymbol
      userId <- mySession.userProfile._id
    } yield profileService.addRecentSymbol(userId, symbol)
  }

  $scope.isRecentSymbol = (symbol: js.UndefOr[String]) => {
    symbol.exists(mySession.isRecentSymbol)
  }

  $scope.removeRecentSymbol = (aSymbol: js.UndefOr[String]) => {
    for {
      symbol <- aSymbol
      userId <- mySession.userProfile._id
    } yield profileService.removeRecentSymbol(userId, symbol)
  }

  ///////////////////////////////////////////////////////////////////////////
  //          Risk Functions
  ///////////////////////////////////////////////////////////////////////////

  $scope.getBetaClass = (beta: js.UndefOr[Double]) => beta map {
    case b if b > 1.3 || b < -1.3 => "volatile_red"
    case b if b >= 0.0 => "volatile_green"
    case b if b < 0 => "volatile_yellow"
    case _ => js.undefined
  }

  $scope.getRiskClass = (riskLevel: js.UndefOr[String]) => riskLevel map {
    case rs if isDefined(rs) && rs.nonEmpty => s"risk_${rs.toLowerCase}"
    case _ => js.undefined
  }

  $scope.getRiskDescription = (riskLevel: js.UndefOr[String]) => riskLevel map {
    case "Low" => "Generally recommended for investment"
    case "Medium" => "Not recommended for inexperienced investors"
    case "High" => "Not recommended for investment"
    case "Unknown" => "The risk level could not be determined"
    case _ => "The risk level could not be determined"
  }

  ///////////////////////////////////////////////////////////////////////////
  //          Market Status Functions
  ///////////////////////////////////////////////////////////////////////////

  $scope.isUSMarketsOpen = () => {
    usMarketStatus match {
      case Left(status) => Option(status.active).orUndefined
      case Right(loading) =>
        if (!loading) {
          usMarketStatus = Right(true)
          console.log("Retrieving market status...")
          marketStatus.getMarketStatus onComplete {
            case Success(status) =>
              // {"stateChanged":false,"active":false,"sysTime":1392092448795,"delay":-49848795,"start":1392042600000,"end":1392066000000}
              // retrieve the delay in milliseconds from the server
              var delay = status.delay
              if (delay < 0) {
                delay = Math.max(status.end - status.sysTime, 300000)
              }

              // set the market status
              console.log(s"US Markets are ${if (status.active) "Open" else "Closed"}; Waiting for $delay msec until next trading start...")
              $scope.$apply(() => usMarketStatus = Left(status))

              // update the status after delay
              console.log(s"Re-loading market status in ${status.delay.minutes}")
              $timeout(() => usMarketStatus = Right(false), 5.minutes)

            case Failure(e) =>
              toaster.error("Failed to retrieve market status")
              console.error(s"Failed to retrieve market status: ${e.getMessage}")
          }
        }
        js.undefined
    }
  }

  ///////////////////////////////////////////////////////////////////////////
  //          Initialization
  ///////////////////////////////////////////////////////////////////////////

  // load the symbol
  if ($scope.q.symbol.isEmpty) {
    // get the symbol
    val symbol = $routeParams.symbol getOrElse $cookies.getOrElse(LastSymbolCookie, mySession.getMostRecentSymbol)

    // load the symbol
    updateQuote(symbol)
  }

  ///////////////////////////////////////////////////////////////////////////
  //          Event Listeners
  ///////////////////////////////////////////////////////////////////////////

  // setup the chart range
  $scope.$watch("options.range", (newValue: js.UndefOr[Any], oldValue: js.Any) => newValue.foreach($cookies.put("chart_range", _)))

}

/**
  * Discover Controller
  * @author Lawrence Daniels <lawrence.daniels@gmail.com>
  */
object DiscoverController {
  val LastSymbolCookie = "QuoteService_lastSymbol"

  def isPerformanceRisk: js.Function1[js.UndefOr[CompleteQuote], Boolean] = (aQuote: js.UndefOr[CompleteQuote]) => aQuote.exists { q =>
    /*q.high52Week.nonEmpty || q.low52Week.nonEmpty || q.change52Week.nonEmpty ||
      q.movingAverage50Day.nonEmpty || q.movingAverage200Day.nonEmpty ||
      q.change52WeekSNP500.nonEmpty || q.beta.nonEmpty*/
    true
  }

  def isIncomeStatement: js.Function1[js.UndefOr[CompleteQuote], Boolean] = (aQuote: js.UndefOr[CompleteQuote]) => aQuote.exists { q =>
    /*q.revenue.nonEmpty || q.revenuePerShare.nonEmpty || q.revenueGrowthQuarterly.nonEmpty ||
      q.grossProfit.nonEmpty || q.EBITDA.nonEmpty || q.netIncomeAvailToCommon.nonEmpty ||
      q.dilutedEPS.nonEmpty || q.earningsGrowthQuarterly.nonEmpty*/
    true
  }

  def isBalanceSheet: js.Function1[js.UndefOr[CompleteQuote], Boolean] = (aQuote: js.UndefOr[CompleteQuote]) => aQuote.exists { q =>
    /*q.totalCash.nonEmpty || q.totalDebt.nonEmpty || q.currentRatio.nonEmpty ||
      q.totalCashPerShare.nonEmpty || q.totalDebtOverEquity.nonEmpty || q.bookValuePerShare.nonEmpty ||
      q.returnOnAssets.nonEmpty || q.profitMargin.nonEmpty || q.mostRecentQuarterDate.nonEmpty ||
      q.returnOnEquity.nonEmpty || q.operatingMargin.nonEmpty || q.fiscalYearEndDate.nonEmpty*/
    true
  }

  def isValuationMeasures: js.Function1[js.UndefOr[CompleteQuote], Boolean] = (aQuote: js.UndefOr[CompleteQuote]) => aQuote.exists { q =>
    /*q.enterpriseValue.nonEmpty || q.trailingPE.nonEmpty || q.forwardPE.nonEmpty ||
      q.pegRatio.nonEmpty || q.priceOverSales.nonEmpty || q.priceOverBookValue.nonEmpty ||
      q.enterpriseValueOverRevenue.nonEmpty || q.enterpriseValueOverEBITDA.nonEmpty ||
      q.operatingCashFlow.nonEmpty || q.leveredFreeCashFlow.nonEmpty*/
    true
  }

  def isShareStatistics: js.Function1[js.UndefOr[CompleteQuote], Boolean] = (aQuote: js.UndefOr[CompleteQuote]) => aQuote.exists { q =>
    /*q.avgVolume3Month.nonEmpty || q.avgVolume10Day.nonEmpty || q.sharesOutstanding.nonEmpty ||
      q.sharesFloat.nonEmpty || q.pctHeldByInsiders.nonEmpty || q.pctHeldByInstitutions.nonEmpty ||
      q.sharesShort.nonEmpty || q.shortRatio.nonEmpty || q.shortPctOfFloat.nonEmpty || q.sharesShortPriorMonth.nonEmpty*/
    true
  }

  def isDividendsSplits: js.Function1[js.UndefOr[CompleteQuote], Boolean] = (aQuote: js.UndefOr[CompleteQuote]) => aQuote.exists { q =>
    /*q.forwardAnnualDividendRate.nonEmpty || q.forwardAnnualDividendYield.nonEmpty ||
      q.trailingAnnualDividendYield.nonEmpty || q.divYield5YearAvg.nonEmpty ||
      q.payoutRatio.nonEmpty || q.dividendDate.nonEmpty || q.exDividendDate.nonEmpty ||
      q.lastSplitFactor.nonEmpty || q.lastSplitDate.nonEmpty*/
    true
  }

  /**
    * Module Expander
    * @author Lawrence Daniels <lawrence.daniels@gmail.com>
    */
  @ScalaJSDefined
  class ModuleExpander(val title: String,
                       val url: String,
                       val icon: String,
                       var expanded: Boolean = false,
                       val visible: js.UndefOr[js.Function] = js.undefined) extends js.Object

}

/**
  * Discover Controller Scope
  * @author Lawrence Daniels <lawrence.daniels@gmail.com>
  */
@js.native
trait DiscoverControllerScope extends AutoCompletionControllerScope {
  // variables
  var expanders: js.Array[ModuleExpander] = js.native
  var options: DiscoverOptions = js.native
  var q: CompleteQuote = js.native
  var ticker: String = js.native
  var tradingHistory: js.Array[js.Object] = js.native

  // functions
  var expandSection: js.Function1[js.UndefOr[ModuleExpander], Unit] = js.native
  var isUSMarketsOpen: js.Function0[js.UndefOr[Boolean]] = js.native
  var loadQuote: js.Function1[js.UndefOr[Any], Unit] = js.native
  var popupNewOrderDialog: js.Function1[js.UndefOr[String], js.UndefOr[js.Promise[NewOrderDialogResult]]] = js.native

  // type-ahead functions
  var formatSearchResult: js.Function1[js.UndefOr[AutoCompleteQuote], js.UndefOr[String]] = js.native
  var onSelectedItem: js.Function3[js.UndefOr[AutoCompleteQuote], js.UndefOr[AutoCompleteQuote], js.UndefOr[String], Unit] = js.native

  // favorite quote functions
  var addFavoriteSymbol: js.Function1[js.UndefOr[String], js.UndefOr[HttpResponse[Profile]]] = js.native
  var isFavorite: js.Function1[js.UndefOr[String], Boolean] = js.native
  var removeFavoriteSymbol: js.Function1[js.UndefOr[String], js.UndefOr[HttpResponse[Profile]]] = js.native

  // recently-viewed quote functions
  var addRecentSymbol: js.Function1[js.UndefOr[String], js.UndefOr[HttpResponse[Profile]]] = js.native
  var isRecentSymbol: js.Function1[js.UndefOr[String], Boolean] = js.native
  var removeRecentSymbol: js.Function1[js.UndefOr[String], js.UndefOr[HttpResponse[Profile]]] = js.native

  // risk functions
  var getBetaClass: js.Function1[js.UndefOr[Double], js.UndefOr[Object]] = js.native
  var getRiskClass: js.Function1[js.UndefOr[String], js.UndefOr[Object]] = js.native
  var getRiskDescription: js.Function1[js.UndefOr[String], js.UndefOr[String]] = js.native

}

/**
  * Discover Options
  * @author Lawrence Daniels <lawrence.daniels@gmail.com>
  */
@ScalaJSDefined
class DiscoverOptions(var range: String) extends js.Object

/**
  * Discover Route Parameters
  * @author Lawrence Daniels <lawrence.daniels@gmail.com>
  */
@js.native
trait DiscoverRouteParams extends js.Object {
  var symbol: js.UndefOr[String] = js.native

}
